src/main/java/org/xerial/snappy/Snappy.java

```diff
@@ -903,4 +903,174 @@ public static String uncompressString(byte[] input, Charset encoding)
+     * @param input
+     * @return bit-shuffled byte array
+     * @throws IOException
+     */
+    public static byte[] bitShuffle(int[] input) throws IOException {
+        byte[] output = new byte[input.length * 4];
+        if (impl.supportBitSuffle()) {
+            impl.bitShuffle(input, 0, 4, input.length * 4, output, 0);
+        } else {
+            arrayCopy(input, 0, input.length * 4, output, 0);
+        }
+        return output;
+    }
+
+    /**
+     * Apply a bit-shuffling filter into the input long array.
+     *
+     * @param input
+     * @return bit-shuffled byte array
+     * @throws IOException
+     */
+    public static byte[] bitShuffle(long[] input) throws IOException {
+        byte[] output = new byte[input.length * 8];
+        if (impl.supportBitSuffle()) {
+            impl.bitShuffle(input, 0, 8, input.length * 8, output, 0);
+        } else {
+            arrayCopy(input, 0, input.length * 8, output, 0);
+        }
+        return output;
+    }
+
+    /**
+     * Apply a bit-shuffling filter into the input float array.
+     *
+     * @param input
+     * @return bit-shuffled byte array
+     * @throws IOException
+     */
+    public static byte[] bitShuffle(float[] input) throws IOException {
+        byte[] output = new byte[input.length * 4];
+        if (impl.supportBitSuffle()) {
+            impl.bitShuffle(input, 0, 4, input.length * 4, output, 0);
+        } else {
+            arrayCopy(input, 0, input.length * 4, output, 0);
+        }
+        return output;
+    }
+
+    /**
+     * Apply a bit-shuffling filter into the input double array.
+     *
+     * @param input
+     * @return bit-shuffled byte array
+     * @throws IOException
+     */
+    public static byte[] bitShuffle(double[] input) throws IOException {
+        byte[] output = new byte[input.length * 8];
+        if (impl.supportBitSuffle()) {
+            impl.bitShuffle(input, 0, 8, input.length * 8, output, 0);
+        } else {
+            arrayCopy(input, 0, input.length * 8, output, 0);
+        }
+        return output;
+    }
+
+    /**
+     * Convert the input bit-shuffled byte array into an original short array.
+     *
+     * @param input
+     * @return a short array
+     * @throws IOException
+     */
+    public static short[] bitUnShuffleShortArray(byte[] input) throws IOException {
+        short[] output = new short[input.length / 2];
+        if (impl.supportBitSuffle()) {
+            impl.bitUnShuffle(input, 0, 2, input.length, output, 0);
+        } else {
+            arrayCopy(input, 0, input.length, output, 0);
+        }
+        return output;
+    }
+
+    /**
+     * Convert the input bit-shuffled byte array into an original int array.
+     *
+     * @param input
+     * @return an int array
+     * @throws IOException
+     */
+    public static int[] bitUnShuffleIntArray(byte[] input) throws IOException {
+        int[] output = new int[input.length / 4];
+        if (impl.supportBitSuffle()) {
```

Comment:
Another concern here is shuffled data is not readable in non x86 platform. This will introduce data incompatibility, which will not be acceptable for snappy-java. 

So using pure-java bitshuffle code is preferable here instead of simply returning non-shuffled data.